src/steamship/agents/context/chathistory.py

```diff
@@ -0,0 +1,161 @@
+                "Multiple ChatHistory objects have been created in this workspace with these context keys."
+            )
+
+    @staticmethod
+    def get_or_create(
+        client: Client,
+        context_keys: Dict[str, str],
+        tags: List[Tag] = None,
+    ) -> ChatHistory:
+
+        file = ChatHistory._get_existing_file(client, context_keys)
+
+        if file is None:
+            tags = tags or []
+            tags.append(Tag(kind=TagKind.DOCUMENT, name=DocTag.CHAT))
+
+            blocks = []
+            file = File.create(
+                client=client,
+                blocks=blocks,
+                tags=tags,
+            )
+        return ChatHistory(file)
+
+    def append_user_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with content provided by the end-user."""
+        tags = tags or []
+        tags.append(
+            Tag(
+                kind=TagKind.CHAT, name=ChatTag.ROLE, value={TagValueKey.STRING_VALUE: RoleTag.USER}
+            )
+        )
+        return self.file.append_block(
+            text=text, tags=tags, content=content, url=url, mime_type=mime_type
+        )
+
+    def append_system_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with content provided by the system, i.e., instructions to the assistant."""
+        tags = tags or []
+        tags.append(
+            Tag(
+                kind=TagKind.CHAT,
+                name=ChatTag.ROLE,
+                value={TagValueKey.STRING_VALUE: RoleTag.SYSTEM},
+            )
+        )
+        return self.file.append_block(
+            text=text, tags=tags, content=content, url=url, mime_type=mime_type
+        )
+
+    def append_agent_message(
+        self,
+        text: str = None,
+        tags: List[Tag] = None,
+        content: Union[str, bytes] = None,
+        url: Optional[str] = None,
+        mime_type: Optional[MimeTypes] = None,
+    ) -> Block:
+        """Append a new block to this with content provided by the agent, i.e., results from the assistant."""
+        tags = tags or []
+        tags.append(
+            Tag(
+                kind=TagKind.CHAT,
+                name=ChatTag.ROLE,
+                value={TagValueKey.STRING_VALUE: RoleTag.ASSISTANT},
+            )
+        )
+        return self.file.append_block(
+            text=text, tags=tags, content=content, url=url, mime_type=mime_type
+        )
+
+    @property
+    def last_user_message(self) -> Optional[Block]:
+        for block in self.file.blocks[::-1]:
+            if block.chat_role == RoleTag.USER:
```

Comment:
it doesn't matter much, but i'd think about refactoring these three methods to use a `last_messages(self, role_tag: str, n: int = 1) -> Optional[Block]` method, so that:

```
def last_system_messages(self, n: int = 1) -> Optional[Block]:
  return self.last_messages(role_tag: RoleTag.SYSTEM, n=n)
```